### PR TITLE
Implement some SkirmishBattleHonors

### PR DESCRIPTION
**Pull request by psykana**

### CODE REVIEW

**Comment by xezon** [comment]
Oh nice. I will get to review today.

File: src/hooker/setupglobals_zh.cpp [hunk 1]
```diff
@@ -775,6 +775,9 @@ GameInfo *&g_theGameInfo = Make_Global<GameInfo *>(PICK_ADDRESS(0x00A2C2B8, 0x04
 SkirmishGameInfo *&g_theSkirmishGameInfo = Make_Global<SkirmishGameInfo *>(PICK_ADDRESS(0x00A2DF10, 0x04CA9F00));
 SkirmishGameInfo *&g_theChallengeGameInfo = Make_Global<SkirmishGameInfo *>(PICK_ADDRESS(0x00A2F374, 0x04CA9E20));
 
+// skirmishbattlehonors.cpp
+int &g_rowsToSkip = Make_Global<int>(PICK_ADDRESS(0x00a2fc68, 0x0098d078));
```
Comment (xezon) [inline src/hooker/setupglobals_zh.cpp]: 0x0098d078 is from Generals or World Builder?

Comment (xezon) [inline src/hooker/setupglobals_zh.cpp]: Prefer uppercase letters in hex, like all other addresses in this file.

File: src/game/common/system/registryget.cpp [hunk 2]
```diff
@@ -90,4 +110,22 @@ bool Get_String_From_Generals_Registry(Utf8String subkey, Utf8String value, Utf8
...
     return false;
 #endif
+}
+
+bool Get_Unsigned_Int_From_Registry(Utf8String subkey, Utf8String value, unsigned int &destination)
+{
+#ifdef PLATFORM_WINDOWS
+    Utf8String key = "SOFTWARE\\Electronic Arts\\EA Games\\Command and Conquer Generals Zero Hour";
+    key += subkey;
+    captainslog_trace("Get_Unsigned_Int_From_Generals_Registry - looking in %s for key %s\n", key.Str(), value.Str());
```
Comment (xezon) [inline src/game/common/system/registryget.cpp]: Can omit \n, however it is the same in the other functions in this file.

File: src/game/common/system/registryget.cpp [hunk 3]
```diff
@@ -39,6 +39,26 @@ static bool getStringFromReg(HKEY hkey, Utf8String subkey, Utf8String key, Utf8S
 
     return false;
 }
+
+static bool getUnsignedIntFromReg(HKEY hkey, Utf8String subkey, Utf8String key, unsigned int &result)
```
Comment (xezon) [inline src/game/common/system/registryget.cpp]: Function name to Thyme guidelines.

File: src/game/common/skirmishbattlehonors.h [hunk 4]
```diff
@@ -22,4 +23,24 @@ class SkirmishBattleHonors : public UserPreferences
...
+    int Get_Losses() const;
+    int Get_China_Campaign_Complete(int level) const;
+    int Get_USA_Campaign_Complete(int level) const;
+    int Get_GLA_Campaign_Complete(int level) const;
+    int Get_Challenge_Campaign_Complete(int general, int level) const;
+    int Get_Honors() const;
 };
+
+void Reset_Battle_Honor_Insertion();
+void Insert_Battle_Honor(GameWindow* window, Image const* image, char earned, int tooltip, int* row, int* column, Utf16String str, int score);
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.h]: `bool earned` perhaps?

File: src/game/common/skirmishbattlehonors.cpp [hunk 5]
```diff
@@ -14,6 +14,18 @@
...
+#include <gadgetlistbox.h>
+#include <gadgetstatictext.h>
+#include <gametext.h>
+#include <gametype.h>
+#include <gamewindowmanager.h>
+#include <image.h>
+#include <maputil.h>
+#include <mouse.h>
+#include <namekeygenerator.h>
+#include <registryget.h>
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: We predominantly use "" for includes.

File: src/game/common/skirmishbattlehonors.cpp [hunk 6]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+
+void Reset_Battle_Honor_Insertion()
+{
+    g_rowsToSkip = 0;
+}
+
+void Insert_Battle_Honor(GameWindow* listbox, Image const* image, char earned, int tooltip, int* row, int* column, Utf16String str, int score)
+{
+    int height = (int)(((float)g_theDisplay->Get_Height() / 600.0) * 41.0);
+    int width = (int)(((float)g_theDisplay->Get_Width() / 800.0) * 40.0);
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: Are these meant to be double?

File: src/game/common/skirmishbattlehonors.cpp [hunk 7]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+    DOMINATION_ONLINE = 1 << 23,
+    STREAK_ONLINE = 1 << 24,
+    CHALLENGE = 1 << 25,
+};
+
+void Populate_Skirmish_Battle_Honors()
+{
+    NameKeyType key = g_theNameKeyGenerator->Name_To_Key("SkirmishGameOptionsMenu.wnd:ListboxInfo");
+    GameWindow *listbox = g_theWindowManager->Win_Get_Window_From_Id(nullptr, key);
+    if (listbox == NULL) {
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: nullptr, several times in this file.

File: src/game/common/skirmishbattlehonors.cpp [hunk 8]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+        return;
+    }
+    SkirmishBattleHonors *BattleHonors = nullptr;
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: Thyme naming convention for in function variables is `battle_honors`

affects several variable names in this file

File: src/game/common/skirmishbattlehonors.cpp [hunk 9]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+    if (window != NULL) {
+        int WinStreak = BattleHonors->Get_Win_Streak();
+        Utf16String str;
+        str.Format(U_CHAR("%d"), WinStreak);
+        Gadget_Static_Text_Set_Text(window, str);
+    }
+
+    key = g_theNameKeyGenerator->Name_To_Key("SkirmishGameOptionsMenu.wnd:StaticTextBestStreakValue");
+    window = g_theWindowManager->Win_Get_Window_From_Id(nullptr, key);
+    if (window != 0) {
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: nullptr, several times in this file.

File: src/game/common/skirmishbattlehonors.cpp [hunk 10]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+    BattleHonors = new SkirmishBattleHonors();
+    listbox->Win_Set_Tooltip_Func(Battle_Honor_Tooltip);
+    Gadget_List_Box_Reset(listbox);
+
+    int honors = BattleHonors->Get_Honors();
+
+    key = g_theNameKeyGenerator->Name_To_Key("SkirmishGameOptionsMenu.wnd:StaticTextStreakValue");
+    GameWindow *window = g_theWindowManager->Win_Get_Window_From_Id(nullptr, key);
+    if (window != NULL) {
+        int WinStreak = BattleHonors->Get_Win_Streak();
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: win_streak

File: src/game/common/skirmishbattlehonors.cpp [hunk 11]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+        Insert_Battle_Honor(listbox, image, 1, CAMPAIGN_USA, &row, &column, 0, 0);
+    } else {
+        image = g_theMappedImageCollection->Find_Image_By_Name("USACampaign_B");
+        Insert_Battle_Honor(listbox, image, (char)(honors >> 0xB) & 1, CAMPAIGN_USA, &row, &column, 0, 0);
+    }
+
+    bool bronze = false;
+    bool silver = false;
+    bool gold = false;
+    for (int i = 0; i < 9; i++) {
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp:269]: I wonder if we have a constant for this number.

File: src/game/common/skirmishbattlehonors.cpp [hunk 12]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+        Insert_Battle_Honor(listbox, image, 1, CAMPAIGN_CHINA, &row, &column, 0, 0);
+    } else if (BattleHonors->Get_China_Campaign_Complete(DIFFICULTY_NORMAL)) {
+        image = g_theMappedImageCollection->Find_Image_By_Name("ChinaCampaign_S");
+        Insert_Battle_Honor(listbox, image, 1, CAMPAIGN_CHINA, &row, &column, 0, 0);
+    } else if (BattleHonors->Get_China_Campaign_Complete(DIFFICULTY_EASY)) {
+        image = g_theMappedImageCollection->Find_Image_By_Name("ChinaCampaign_B");
+        Insert_Battle_Honor(listbox, image, 1, CAMPAIGN_CHINA, &row, &column, 0, 0);
+    } else {
+        image = g_theMappedImageCollection->Find_Image_By_Name("ChinaCampaign_B");
+        Insert_Battle_Honor(listbox, image, (char)(honors >> 0xC) & 1, CAMPAIGN_CHINA, &row, &column, 0, 0);
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: I think this is meant to be written as

(honors & CAMPAIGN_CHINA) != 0

Several times in this file.

File: src/game/common/skirmishbattlehonors.cpp [hunk 13]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+        image = g_theMappedImageCollection->Find_Image_By_Name("Endurance_B");
+        Insert_Battle_Honor(listbox, image, 0, ENDURANCE, &row, &column, 0, 0);
+    }
+
+    // Apocalypse: build all superweapons
+    image = g_theMappedImageCollection->Find_Image_By_Name("Apocalypse");
+    Insert_Battle_Honor(listbox, image, (char)(honors >> 0x11) & 1, APOCALYPSE, &row, &column, 0, 0);
+
+    // Blitz: win in (x) minutes or less
+    if ((honors & 0x4000) != 0) {
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: BLITZ_5

File: src/game/common/skirmishbattlehonors.cpp [hunk 14]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+        image = g_theMappedImageCollection->Find_Image_By_Name("HonorBlitz5");
+        Insert_Battle_Honor(listbox, image, 1, BLITZ_5, &row, &column, 0, 0);
+    } else if ((honors & 0x8000u) == 0) {
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: BLITZ_10

Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: Perhaps flip condition to `(honors & BLITZ_10) != 0` and list the enabled image first, and disabled image last, like in the other cases.

File: src/game/common/skirmishbattlehonors.cpp [hunk 15]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+    } else if (wins >= 100) {
+        image = g_theMappedImageCollection->Find_Image_By_Name("Domination_100");
+        Insert_Battle_Honor(listbox, image, 1, DOMINATION, &row, &column, 0, wins);
+    } else {
+        image = g_theMappedImageCollection->Find_Image_By_Name("Domination_100");
+        Insert_Battle_Honor(listbox, image, 0, DOMINATION, &row, &column, 0, wins);
+    }
+
+    // Ultimate: beat every map vs max number of hard opponents
+    char earned = 1;
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: bool, true

File: src/game/common/skirmishbattlehonors.cpp [hunk 16]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+    for (auto it = g_theMapCache->begin(); it != g_theMapCache->end(); it++) {
+        if (it->second.m_isOfficial && it->second.m_isMultiplayer) {
+            int numBeat = BattleHonors->Get_Endurance_Medal(it->first, 4);
+            if (numBeat < it->second.m_numPlayers - 1) {
+                earned = 0;
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: false

File: src/game/common/skirmishbattlehonors.cpp [hunk 17]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+        Insert_Battle_Honor(listbox, image, 1, DOMINATION, &row, &column, 0, wins);
+    } else if (wins >= 500) {
+        image = g_theMappedImageCollection->Find_Image_By_Name("Domination_500");
+        Insert_Battle_Honor(listbox, image, 1, DOMINATION, &row, &column, 0, wins);
+    } else if (wins >= 100) {
+        image = g_theMappedImageCollection->Find_Image_By_Name("Domination_100");
+        Insert_Battle_Honor(listbox, image, 1, DOMINATION, &row, &column, 0, wins);
+    } else {
+        image = g_theMappedImageCollection->Find_Image_By_Name("Domination_100");
+        Insert_Battle_Honor(listbox, image, 0, DOMINATION, &row, &column, 0, wins);
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: When changing `char` to `bool`, please change third argument to true/false accordingly in all calls.

File: src/game/common/skirmishbattlehonors.cpp [hunk 18]
```diff
@@ -27,3 +39,617 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+        str = g_theGameText->Fetch("TOOLTIP:BattleHonors", nullptr);
+        g_theMouse->Set_Cursor_Tooltip(str, -1, 0, 1.0f);
+    } else {
+        honor = (int)Gadget_List_Box_Get_Item_Data(listbox, row, column);
+        score = (int)Gadget_List_Box_Get_Item_Data(listbox, row - 1, column);
+        if (honor == 0) {
+            str = g_theGameText->Fetch("TOOLTIP:BattleHonors", nullptr);
+            g_theMouse->Set_Cursor_Tooltip(str, -1, 0, 1.0f);
+        } else if (honor != 0) {
+            if (honor & 0x8000000) { // not earned
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: Is there perhaps an enum value missing for 1<<27 in SkirmishBattleHonorsTooltips?

File: src/hooker/setupglobals_zh.cpp (hunk 1 above)
Comment (psykana) [reply src/hooker/setupglobals_zh.cpp]: World Builder

File: src/game/common/skirmishbattlehonors.cpp (hunk 6 above)
Comment (psykana) [reply src/game/common/skirmishbattlehonors.cpp:109]: You're right, `FILD` instruction is used in ZH

File: src/game/common/skirmishbattlehonors.cpp (hunk 12 above)
Comment (psykana) [reply src/game/common/skirmishbattlehonors.cpp]: Didn't realize they were the same with three different bases all over

File: src/game/common/skirmishbattlehonors.cpp (hunk 11 above)
Comment (psykana) [reply src/game/common/skirmishbattlehonors.cpp:269]: I don't see anything related. Each challenge campaign general has either 7 or 8 missions.

**Comment by xezon** [comment]
Code wise looks good to me. I will take a look at it ingame later today.

**Comment by xezon** [comment]
Like breaks in tool tips are wrong.

## Original and Before this change

shot_20240325_183112_1

## This change

shot_20240325_183309_1

**Review by xezon — COMMENTED** [review]
See above

**Comment by psykana** [comment]
I tested before/after at different resolutions so that went over my head, fixed.

**Review by xezon — APPROVED** [review]
Looks good.

**Comment by psykana** [comment]
Can I run format check locally so we don't have to drag this any more?

**Review by xezon — COMMENTED** [review]
Formatting check is failing. You need to apply clang format, with version 10.0.0 (as part of LLVM). If you are using Visual Studio, you can refer to custom clang-format exe in settings page, otherwise it uses clang-format as part of Visual Studio distribution.

**Comment by xezon** [comment]
```
/home/runner/work/Thyme/Thyme/src/game/common/skirmishbattlehonors.cpp:379:9: error: no matching member function for call to 'Format'
    str.Format(L"%10d", wins);
    ~~~~^~~~~~
/home/runner/work/Thyme/Thyme/src/game/common/system/unicodestring.h:140:10: note: candidate function not viable: no known conversion from 'const wchar_t [5]' to 'const unichar_t *' (aka 'const char16_t *') for 1st argument
    void Format(const unichar_t *format, ...);
         ^
/home/runner/work/Thyme/Thyme/src/game/common/system/unicodestring.h:141:10: note: candidate function not viable: no known conversion from 'const wchar_t [5]' to 'Utf16String' for 1st argument
    void Format(Utf16String format, ...);
         ^
/home/runner/work/Thyme/Thyme/src/game/common/skirmishbattlehonors.cpp:430:17: error: cast from pointer to smaller type 'int' loses information
        honor = (int)Gadget_List_Box_Get_Item_Data(listbox, row, column);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/Thyme/Thyme/src/game/common/skirmishbattlehonors.cpp:431:17: error: cast from pointer to smaller type 'int' loses information
        score = (int)Gadget_List_Box_Get_Item_Data(listbox, row - 1, column);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4 warnings and 3 errors generated.
```

File: src/game/common/skirmishbattlehonors.cpp [hunk 19]
```diff
@@ -27,3 +39,619 @@ int SkirmishBattleHonors::Get_Endurance_Medal(Utf8String name, int level) const
...
+                } else if (honor & DOMINATION_ONLINE) {
+                    str = g_theGameText->Fetch("TOOLTIP:BattleHonorDominationOnlineDisabled", nullptr);
+                    g_theMouse->Set_Cursor_Tooltip(str, -1, nullptr, 1.5f);
+                }
+            }
+            if (honor & LOYALTY_USA) {
+                str = g_theGameText->Fetch("TOOLTIP:BattleHonorLoyaltyUSA", nullptr);
+                g_theMouse->Set_Cursor_Tooltip(str, -1, nullptr, 1.5f);
+            }
+            if (honor & LOYALTY_CHINA) {
```
Comment (xezon) [inline src/game/common/skirmishbattlehonors.cpp]: Should this be `else if`? And below?

Comment (psykana) [reply src/game/common/skirmishbattlehonors.cpp]: You're right. Should be good to go now.

**Pull request merged**